Add GA API policy and describe feature gates

dibyom opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes


[Add GA API policy and describe feature gates](https://github.com/tektoncd/pipeline/pull/5634/commits/3a91ae83b56c0470ff501fcb7dbc65493d7457b7)

This commit describes our API policiy for our GA APIs (e.g. v1). It also
clarifies how TEP-33 Feature Gates interacts with CRD API versions.

Fixes https://github.com/tektoncd/pipeline/issues/5633
# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Updates API compatibility policy for the V1 api version
```

dibyom commented:
/kind misc

JeromeJu commented on api_compatibility_policy.md:
```diff
@@ -43,21 +40,41 @@ alpha:
+
+### Beta
+
+- The `apiVersion` field of the CRD contains contain `beta` (for example, `v1beta1`).
```
nit: we might want one less contain here

dibyom replied:
indeed

JeromeJu commented on api_compatibility_policy.md:
```diff
@@ -43,21 +40,41 @@ alpha:
+
+- For Alpha CRDs, the `apiVersion` contains the `alpha` (e.g. `v1alpha1`)
+  
+- Alpha feature in beta or stable CRDs are disabled by default and must be enabled by [setting `enable-api-fields` to `alpha`](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features)
```
Wondering if it make sense and need to mention "alpha" feature is a superset of "beta"?

dibyom replied:
do you mean if you enable alpha, beta features are also enabled?

lbernick commented:
@dibyom is this meant to clarify the existing api policy, or to propose the api policy discussed in this week's WG and your [API versioning doc](https://docs.google.com/document/d/15vD5vbK5UKCDFMoOqOwEc_NYACthtCZJlTJiiEEoljY)? (or is this just not ready for review yet? :) )

dibyom commented:
@tektoncd/core-maintainers PTAL!

lbernick commented on api_compatibility_policy.md:
```diff
@@ -43,21 +40,83 @@ alpha:
+
+- Alpha feature in beta or GA CRDs are disabled by default and must be enabled by [setting `enable-api-fields` to `alpha`](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features)
+
+- These features may be dropped or backawards incompatible changes made at any time though will be given at least one release worth of warning
```
```suggestion
- These features may be dropped or backwards incompatible changes made at any time though will be given at least one release worth of warning
```

dibyom commented on api_compatibility_policy.md:
```diff
@@ -43,21 +40,83 @@ alpha:
+
+- We will not make any backwards incompatible changes to fields that are stable without incrementing the api version.
+
+- Stable APIs do not follow the [Go libraries compatibility policy](#go-libraries-compatibility-policy) i.e. backwards incompatible changes to the Go struct is not allowed.
```
Hmm, I was thinking of only a subset of the go packages - esp. the ones in pkg/api/v1 and pkg/client. I'm not sure we need to have all exported functions under this

lbernick replied:
Honestly maybe everything that's not in apis should be in an internal package?

vdemeester replied:
> Honestly maybe everything that's not in apis should be in an internal package?

In theory yes 😬. Like in theory most of what is in `pkg` today should be in `internal/pkg` or something unless required otherwise (like `pkg/apis`, the generated clients, …). In practice using `internal` is not always super known by go dev.

wlynch replied:
@dibyom I'd say anything visible in https://pkg.go.dev/github.com/tektoncd/pipeline is fair game.

+1 for moving things to internal.

lbernick replied:
Any reason not to move things to internal? Vincent it sounds like you're saying we haven't done this just because most people aren't aware of the option? If that's the case I'm happy to make this change

lbernick replied:
tracked in https://github.com/tektoncd/pipeline/issues/5679.

vdemeester replied:
There is more to it than just "move" though. We did not use it too much initially, because it's not that common to use in the OSS golang community. But now, the project is there for a couple of years already, there is probably quite some project depending on some parts of our go packages, even only in the `tektoncd` org (triggers, …). We need to make sure we are not doing a "switch off" on those and, for each package, look at what is exported, why, is it useful, and is it used elsewhere ?

dibyom replied:
At least for triggers, we only use pkg/client and pkg/apis (we have one test relying on pkg/names but that can probably be fixed easily). I do agree with your statement @vdemeester. My initial suggestion was just keeping only pkg/apis and pkg/client under the API policy - I don't think its feasible for us to promise no breaking changes to all of pkg/

lbernick replied:
I agree with Billy that things indexed on pkg.go.dev make sense to cover in our compatibility policy. I would find it confusing to have a compatibility policy that covers some of our public go packages but not others. I agree we don't want to promise no changes to pkg/, but we could say we won't remove/rename a deprecated function/const for x time?

dibyom replied:
> we could say we won't remove/rename a deprecated function/const for x time?

We could perhaps provide that but backwards compatibility technically also includes things like adding new arguments or return values to a function: https://go.dev/blog/module-compatibility 

@wlynch I'm not sure how we can both guarantee go module compatibility while also making rapid changes to alpha APIs if all of pkg/ is under scope.

I think clearly stating out which packages are in scope and which are not is a decent middle ground. Over time, we could move the the out of scope packages into internal. Another option would be to clearly write out what subset of compatibility we can provide e.g. we can support not deprecating/removing functions but we'd probably want the ability modify function signatures in most of pkg/

afrittoli replied:
I'm a bit confused about how the policy can be different for stable APIs only - as go libraries are not necessarily bound to a specific API version. 

Does this summary make sense:
- go APIs specific to stable Tekton APIs are subject to the new policy, i.e. we cannot do backwards incompatible changes. Also, a future `v2` public API must be separated in code from the `v1` public API
- go APIs specific to alpha/beta Tekton APIs can be subject to backwards-incompatible changes
- go APIs which are shared between alpha/beta and stable that needs to stay public would have to be separated,  one copy under the stable tree, one copy under the alpha/beta one
- go APIs that we don't need to be public would have to become private, and we could (1) move them to private (2) create a public shim (3) deprecate the shim right away. This would have to happen before v1

dibyom replied:
The first two points make sense to me. 

- For public go apis shared between alpha and beta - why do we need to copy unless we are making a breaking change? There is quite a bit of complexity involved in this so I'm sure we should do this unless we really need to

- I support moving them to private incrementally but I don't think we need to block the v1 API release on this. I'm unsure why we'd need to deprecate the shim.

We had some discussion on this in the last API WG - the conclusion was that I'd update the policy to mention add  guidelines for making changes to the go libraries. Specifically this would be to avoid making backwards incompatible changes to the go APIs pertaining to the V1 API version.

afrittoli replied:
> The first two points make sense to me.
> 
> * For public go apis shared between alpha and beta - why do we need to copy unless we are making a breaking change? There is quite a bit of complexity involved in this so I'm sure we should do this unless we really need to

I meant shared between alpha/beta and stable. If we can make sure (through testing) that we would catch those cases, then it would be fine to not copy the code until a breaking change is needed on alpha or beta side.

> * I support moving them to private incrementally but I don't think we need to block the v1 API release on this. I'm unsure why we'd need to deprecate the shim.

We don't need to block v1 only on this as long as our API stability policy excludes them upfront. 
If we create a shim, it would only be to give existing users more time to move away from those APIs that we want to become private. But we would have to deprecate the shim to signal our users that that is only a temporary solution.
Alternatively we could deprecate the API before making it private, and when it moves to private provide no shim at all.

> 
> We had some discussion on this in the last API WG - the conclusion was that I'd update the policy to mention add guidelines for making changes to the go libraries. Specifically this would be to avoid making backwards incompatible changes to the go APIs pertaining to the V1 API version.

dibyom commented:
I updated and split up the PR to separate out the bit around go libraries compatibility since that seems to be more controversial. Happy to move forward without the changes to Go libraries if that is easier! @vdemeester @lbernick Let me know if the update captures the discussion and what you had in mind

Also, @afrittoli and any other @tektoncd/core-maintainers this should be ready for another review!

lbernick commented on api_compatibility_policy.md:
```diff
@@ -29,35 +25,96 @@ A backwards incompatible change would be a change that requires a user to update
+
+- The `apiVersion` field of the CRD is `vX` where `X` is an integer.
+
+- Stable API versions remain available for all future releases within a Tekton major version.
```
does "tekton major version" here refer to semver?

dibyom replied:
yes - will update the wording to include semver major version

lbernick commented on api_compatibility_policy.md:
```diff
@@ -29,35 +25,96 @@ A backwards incompatible change would be a change that requires a user to update
-## Alpha, Beta and GA
+## CRD API Versions
+
+The `apiVersion` field in a Tekton CRD determines whether the overall API (and its default behaviors) is considered to be in `alpha`, `beta`, or `GA`. The use of `alpha`, `beta` and `GA` (aka `stable`) follows the corresponding Kubernetes [API stages definition](https://kubernetes.io/docs/reference/using-api/#api-versioning) and its [deprecation policy](https://kubernetes.io/docs/reference/using- ...
```
I think saying we follow k8s deprecation policy is a bit confusing actually (the main differences I see are the ability to mark a beta api as deprecated immediately, rather than waiting 9 months in addition to the 9 months required to remove it) and our feature flags policy.

dibyom replied:
ok, yeah the k8s deprecation policy also includes things like cli flags that are not relevant to us at the moment

lbernick commented on api_compatibility_policy.md:
```diff
@@ -29,35 +25,96 @@ A backwards incompatible change would be a change that requires a user to update
+
+- Alpha and Beta features may be present within a stable API version. However, they will not be enabled by default and must be enabled by setting `enable-api-fields` to `alpha` or `beta`.
+
+## Features Gates
```
```suggestion
## Feature Gates
```

lbernick commented on api_compatibility_policy.md:
```diff
@@ -29,35 +25,96 @@ A backwards incompatible change would be a change that requires a user to update
+- Beta features may be deprecated or changed in a backwards incompatible way by following the same process as [Beta CRDs](#beta-crds) 
+  i.e. by providing a 9 month support period.
+
+- Beta features are reviewed for promotion to beta at a GA/Stable basis. However, there is no guarantee that they will be promoted to GA/stable.
```
```suggestion
- Beta features are reviewed for promotion to GA/Stable at a regular basis. However, there is no guarantee that they will be promoted to GA/stable.
```

afrittoli commented on api_compatibility_policy.md:
```diff
@@ -89,9 +145,16 @@ These changes must be approved by [more than half of the project OWNERS](OWNERS)
+
+- Changing or removing a field in a go struct in `pkg/apis/pipeline/v1alpha1` (or another alpha API) is not considered a backwards incompatible change.
+
+- Renaming a field in `pkg/apis/pipeline/v1beta1` (or another beta API) is not considered a backwards incompatible change. However, removing a field is considered a backwards incompatible change and requires a major version bump.
```
Similar comment as for alpha. We cannot really define what is and is not backwards compatible.
The [reference](https://go.dev/blog/module-compatibility) you linked below looks like a good one to explain what is and what is not backwards compatible. 

I believe what should change between alpha and beta is the amount of notice we give users for backward incompatible changes. Additionally, we can define changes that require a new major beta version no matter the amount of notice - for instance, we can say that removing a field requires a new major version.

It seems to me that separating renaming from removing is artificial. Renaming is at all effects removal of the old name + the addition of the new name. While the same content is stored, old clients won't find the new field without changes.

I think the way to do renaming is to provide the new name and keep the old one but deprecate it. 
After the agreed deprecation period is met, we remove the old field.

In stable APIs instead, we can never remove a field. 
We could add the new name, which is backwards compatible, but only a new stable version could remove the field

lbernick replied:
I agree that separating removing from renaming is a bit artificial (even though it was partially my idea 😅 ) but I'm not sure how realistic it is to announce to users in advance a change we want to make to the go libs. (maybe it is?)
Do you think it would be OK to just allow backwards incompatible changes in v1beta1 go libs? since the definition of beta is that features won't be removed, but the details may change-- I think if we're not changing the yaml/json definitions it's a reasonable policy.

afrittoli replied:
I think it's ok to allow backwards incompatible changes in the beta API, but I think we should provide notice.
Similar to the API policy:
- alpha we can do backwards incompatible without notice 
- beta we can do backwards incompatible changes with notice (i.e. going through deprecation)
- stable we cannot do backwards incompatible changes, they have to go in a new version

dibyom replied:
I tend to agree...what I'm a bit worried about is how well we can actually implement providing notice for the beta  changes - things like adding an extra arg or a return value to a function are not backwards incompatible. Do you see us providing notice for changes like those across all exported functions in pkg/? If so, how long would the notice period be? (Personally, I'd be ok with a `action-required` release note)

afrittoli replied:
In terms of how to deprecate, I think action required release notes are good, and comment in the code, so that's shown by IDEs 

In terms of duration, what about three releases? I think our current API policy of 9 months is too much, especially it will be once we have a v1 out, so definitely I would go shorter than that.

dibyom replied:
updated to 3 releases

afrittoli reviewed (commented):
Thank you for this PR - see some comments on the go library side.

lbernick commented on api_compatibility_policy.md:
```diff
@@ -29,35 +25,95 @@ A backwards incompatible change would be a change that requires a user to update
+
+- GA/Stable features are enabled by default
+
+- GA/Stable features will not be removed or changed in a backwards incompatible manner.
```
```suggestion
- GA/Stable features will not be removed or changed in a backwards incompatible manner without incrementing the API version
```

vdemeester commented on api_compatibility_policy.md:
```diff
@@ -89,9 +145,16 @@ These changes must be approved by [more than half of the project OWNERS](OWNERS)
+
+- Backwards incompatible changes in `pkg/apis/pipeline/v1beta1` (or another beta API) require providing a notice of the upcoming change for at least 3 releases before making the change. However, they do not require a major version bump.
+
+- For `pkg/apis/pipeline/v1` (or any other stable APIs), any [backwards incompatible change](https://go.dev/blog/module-compatibility) will require a major version bump.
```
By *major* we mean `{major}.{minor}.{bugfix}` ?

lbernick replied:
I think this was intended to refer to semver since it's talking about the go libraries-- @dibyom can you update to clarify?

dibyom replied:
I was referring to semver. I do think we need to bump the API version if we make a breaking change to the API itself, but if we are just renaming a function or something, I think we can just update the semver version.

afrittoli replied:
I'm confused. We have the API version, which is not semver, and the version of the go module, for which we only have used minor and patch, major is still 0 (e.g. 0.41.1).  So we can do a breaking change in the code in `pkg/apis/pipeline/v1`, as long as we release a v1.0.0, v2.0.0 etc?

About renaming, for an API in a package associated to a stable version, I would prefer we never did it in one go, regardless of the version. We should create a new field / function with the name we want and deprecate the old one.

dibyom replied:
> So we can do a breaking change in the code in pkg/apis/pipeline/v1, as long as we release a v1.0.0, v2.0.0 etc?

Technically yes, the intent is to strongly discourage these kinds of changes since major version bumps should be rare. (If it changes the user facing API and not just a function, we'd need to change the API version as well)

> About renaming, for an API in a package associated to a stable version, I would prefer we never did it in one go, regardless of the version. We should create a new field / function with the name we want and deprecate the old one.

I agree, I don't think anything here is saying we should not do that. (I can include that explicitly if that helps)

dibyom commented:
/hold - need to fix a commit

dibyom commented:
/hold cancel

dibyom commented:
[Splitting off the discussion around Go libraries and releases to a separate PR]

vdemeester commented:
/approve

afrittoli reviewed (commented):
Thank you!
/approve

lbernick commented:
/lgtm
